scripts/tests/calibnet_rpc_checks.sh

```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env bash
+# This script is used to test both online and offline RPC API server against lotus.
+# It's run in CI, and uses forest-tool api compare subcommand to test RPC endpoints.
+
+set -euxo pipefail
+
+pushd "$( cd "$(dirname "${BASH_SOURCE[0]}")" ; pwd -P )"
+LOTUS_IMAGE="filecoin/lotus-all-in-one:v1.26.0-rc3-calibnet"
+FOREST_PATH="forest"
+FOREST_CLI_PATH="forest-cli"
+FOREST_TOOL_PATH="forest-tool"
+PORTS=(1234 2345 2346)
+
+# Function to stop services on specified ports
+stop_services() {
+    kill -KILL "$FOREST_PID" || true
+    kill -KILL "$FOREST_TOOL_PID" || true
+    kill -KILL "$LOTUS_PID" || true
+    docker container rm -f lotus || true
+    for port in "${PORTS[@]}"; do
+        fuser -k "$port/tcp" || true
+    done
+}
+trap stop_services EXIT
+
+# Run forest daemon RPC server at 2345
+"$FOREST_PATH" --chain calibnet --encrypt-keystore false --no-gc --height=-900 --auto-download-snapshot &
+FOREST_PID=$!
+
+"$FOREST_CLI_PATH" sync wait
+EXPORTED_SNAPSHOT="latest.forest.car.zst"
+"$FOREST_CLI_PATH" snapshot export -o "$EXPORTED_SNAPSHOT" -d=900 --include-message-receipts
+
+old_snapshot=forest_diff_calibnet_2022-11-02_height_0+3000.forest.car.zst
+curl --location --remote-name "https://forest-archive.chainsafe.dev/calibnet/diff/$old_snapshot"
+
+# Run forest-tool RPC server at 2346
+"$FOREST_TOOL_PATH" api serve "$EXPORTED_SNAPSHOT" "$old_snapshot" --chain calibnet --port 2346 &
+FOREST_TOOL_PID=$!
+
+# Run lotus daemon RPC server at 1234
+docker run --name lotus --rm \
+  -e LOTUS_API_LISTENADDRESS=/ip4/0.0.0.0/tcp/1234/http \
+  -e LOTUS_FEVM_ENABLEETHRPC=1 \
+  -e LOTUS_CHAINSTORE_ENABLESPLITSTORE=false \
+  -p 1234:1234 \
+  -v "$PWD:/pwd" \
+  -v /var/tmp/filecoin-proof-parameters:/var/tmp/filecoin-proof-parameters \
+  "$LOTUS_IMAGE" lotus daemon --remove-existing-chain --import-snapshot "/pwd/$EXPORTED_SNAPSHOT" &
+LOTUS_PID=$!
+
+# Check if services on ports have started
+for port in "${PORTS[@]}"; do
+    until nc -z localhost "$port"; do
+        sleep 10
+    done
+done
+
+until docker exec lotus lotus sync wait; do
+    sleep 5
+done
+
+# Assert an old block is present, given that the old snapshot is used.
+# https://calibration.filfox.info/en/block/bafy2bzacecpjvcld56dazyukvj35uzwvlh3tb4ga2lvbgbiua3mgbqaz45hbm
+temp_dir=$(mktemp -d)
+FULLNODE_API_INFO="/ip4/127.0.0.1/tcp/2346/http" forest-cli chain block -c bafy2bzacecpjvcld56dazyukvj35uzwvlh3tb4ga2lvbgbiua3mgbqaz45hbm | jq . > "$temp_dir/block.json"
+
+# assert block is as expected
+diff "test_data/calibnet_block_3000.json" "$temp_dir/block.json"
+
+# Compare the http endpoints
+# forest daemon vs lotus daemon
```

Comment:
@LesnyRumcajs any objections to moving RPC tests to this script?

Reply:
What's the improvement? In general, I'd rather we have a dockerized environment for tests so that anyone can run it anywhere without latest Forest binaries present or certain host ports available.

Also, this hardcodes in the script the Lotus image tag, which now is automatically updated via https://github.com/ChainSafe/forest/blob/main/.github/workflows/lotus-api-bump.yml

Reply:
@LesnyRumcajs as a CI test, I found it inconvenient to not be able to run for a draft PR, also it's more expensive to rely on buildjet

Reply:
> as a CI test, I found it inconvenient to not be able to run for a draft PR, also it's more expensive to rely on buildjet. 

You can always comment out `if: ...` in the workflow to trigger them in draft PRs (or we can have a dispatch trigger to have it run manually, on demand).

> When troubleshooting issues, I need to run it multiple times (and better to have the PR in draft state)

Why do you need to run it multiple times in the CI? Having it dockerized makes it more environment-resistant, so you can easily run it locally.

> I think we could have another dockerized test after pushing forest images (or on daily basis), using only officially published images. But using a fully dockerized CI test for dev builds (where it's likely to fail and need iterations) is over complicated for the goal IMHO

Dev builds failing should be resolved locally, and fully dockerized tests make it easy to do so on any machine.

All in all, I disagree on un-dockerizing this; I find it significantly less complicated and easier to reason about when all is done in terms of containers. That said, if the rest of the team feels otherwise (or at least two other reviewers), then I'm not going to block the change with my vision. If you decide to do so, please also update the automatic Lotus version bump workflow.

Reply:
@LesnyRumcajs How about a docker-compose setup like devnet tests? That way it does not depend on the full docker build step (which I think is some extra overhead for local develepment)

Reply:
In local development, you only need `docker build -t ghcr.io/chainsafe/forest:edge-fat .`, then in `scripts/tests/api_compare` run `./setup.sh && ./api_compare.sh`. Not sure what is the overhead for local development there.

In a CI, sure, we could attach ourselves to https://github.com/ChainSafe/forest/blob/main/scripts/devnet/forest_ci.dockerfile instead of relying on the Docker workflow artefact. I assumed folks prefer running tests on their local machines; it makes sense to not interfere with their workflow.

Reply:
docker build is pretty slow :) guess my laptop is far less powerful than your dev box. I prefer the way how devnet test minimizes the docker build time by copying local binaries

Reply:
`docker build` should have the exact same speed as a local build, including the incremental changes. The only slow part is the initial build, which will do it from scratch.

Seems it makes sense to support both approaches?

Reply:
> Seems it makes sense to support both approaches?

@LesnyRumcajs I think the devnet test supports both, and do you think I can change the RPC test to be similar, I'd really love to see RPC test running for draft PRs without commenting out `if: ` line or invoking buildjet build agents.

```yaml
  forest:
    user: root
    depends_on:
      lotus_config:
        condition: service_completed_successfully
    build:
      context: ../../.
      dockerfile: ${FOREST_DOCKERFILE_OVERRIDE:-Dockerfile}
```

Reply:
Yeah, as long as all the services end up running in a containerized environment, I'm good. Devnet has a pretty good design (_patting myself on the back_) that we can re-use.

> P.S. I think we should remove the below lines from the script and set it only in CI, that way people can run it locally without modification

No strong opinion, I usually do `docker compose up` manually anyway.

Reply:
> Yeah, as long as all the services end up running in a containerized environment, I'm good. 

@LesnyRumcajs Done.